src/connectors/impls/gpubsub/gsub.rs

```diff
@@ -0,0 +1,331 @@
+        let skip_authentication = self.config.skip_authentication;
+
+        let connect_to_pubsub = move || -> Result<PubSubClient> {
+            if skip_authentication {
+                info!("Skipping auth...");
+                return Ok(SubscriberClient::with_interceptor(
+                    channel.clone(),
+                    AuthInterceptor {
+                        token: Box::new(|| Ok(Arc::new(String::new()))),
+                    },
+                ));
+            }
+
+            let token = Token::new()?;
+
+            Ok(SubscriberClient::with_interceptor(
+                channel.clone(),
+                AuthInterceptor {
+                    token: Box::new(move || match token.header_value() {
+                        Ok(val) => Ok(val),
+                        Err(_) => Err(Status::unavailable(
+                            "Failed to retrieve authentication token.",
+                        )),
+                    }),
+                },
+            ))
+        };
+
+        if let Some(task_handle) = self.task_handle.take() {
+            task_handle.cancel().await;
+        }
+
+        let client = connect_to_pubsub()?;
+
+        let client_background = connect_to_pubsub()?;
+
+        let (tx, rx) = async_std::channel::bounded(QSIZE.load(Ordering::Relaxed));
+        let (ack_tx, ack_rx) = async_std::channel::bounded(QSIZE.load(Ordering::Relaxed));
+
+        let join_handle = async_std::task::spawn(consumer_task(
+            client_background,
+            tx,
+            self.ack_ids.clone(),
+            self.config.subscription_id.clone(),
+            Duration::from_nanos(self.config.ack_deadline),
+            ack_rx,
+        ));
+
+        self.receiver = Some(rx);
+        self.ack_sender = Some(ack_tx);
+        self.client = Some(client);
+        self.task_handle = Some(join_handle);
+
+        Ok(true)
+    }
+
+    async fn pull_data(&mut self, pull_id: &mut u64, ctx: &SourceContext) -> Result<SourceReply> {
+        let receiver = self
+            .receiver
+            .as_mut()
+            .ok_or(ErrorKind::PubSubClientNotAvailable(
+                "The receiver is not connected",
+            ))?;
+        let (ack_id, data) = match receiver.recv().await? {
+            Ok(response) => response,
+            Err(error) => {
+                let error_kind = error.kind();
+                return if let ErrorKind::Timeout(_) = error_kind {
+                    ctx.swallow_err(
+                        ctx.notifier.connection_lost().await,
+                        "Failed to notify about PubSub connection loss",
+                    );
+
+                    Ok(SourceReply::StreamFail(DEFAULT_STREAM_ID))
+                } else {
+                    Err(error)
+                };
+            }
+        };
+
+        *pull_id = ack_id;
+
+        Ok(SourceReply::Data {
+            origin_uri: EventOriginUri::default(),
+            data,
+            meta: None,
```

Comment:
It would be super nice to have some metadata with the message `id`, the `ordering_key`, `publish_time` and the attributed from the [`PubsubMessage`](https://github.com/mechiru/googapis/blob/v0.6.0/googapis/genproto/google.pubsub.v1.rs#L409).

Similar to how it is done in kafka: https://github.com/tremor-rs/tremor-runtime/blob/main/src/connectors/impls/kafka/consumer.rs#L343